src/main/java/dev/ecattez/shahmat/board/ChessGame.java

```diff
@@ -0,0 +1,129 @@
+package dev.ecattez.shahmat.board;
+
+import dev.ecattez.shahmat.event.TradeProposed;
+import dev.ecattez.shahmat.board.pawn.PawnTradingRankVisitor;
+import dev.ecattez.shahmat.command.MovePiece;
+import dev.ecattez.shahmat.command.TradePawn;
+import dev.ecattez.shahmat.event.BoardEvent;
+import dev.ecattez.shahmat.event.PawnTraded;
+import dev.ecattez.shahmat.event.PieceCaptured;
+import dev.ecattez.shahmat.event.PieceMoved;
+import dev.ecattez.shahmat.event.PiecePositioned;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+public class ChessGame {
+
+    private static final PieceFactory PIECE_FACTORY = new PieceBox(); // fixme: change it
+
+    private static final PieceTypeVisitor<MovingStrategy> MOVING_STRATEGIES = new MovingRules();
+
+    private static final PieceColorVisitor<Integer> TRADING_RANKS = new PawnTradingRankVisitor();
+
+    public static List<BoardEvent> move(List<BoardEvent> history, MovePiece command) throws InvalidMove {
+        Square from = command.from;
+        Square to = command.to;
+
+        if (from.equals(to)) {
+            throw new InvalidMove("Current position and destination are the same square");
+        }
+
+        List<BoardEvent> events = new LinkedList<>();
+        Board board = replay(history);
+        Piece piece = command.piece;
+
+        board.getPiece(from)
+            .filter(pieceOnBoard ->
+                pieceOnBoard.equals(piece)
+                    && canBeMoved(pieceOnBoard, from, to, board)
```

Comment:
Si la pièce ne peut pas être déplacée, ne faut-il pas renvoyer une erreur du type `ImpossibleToMove(piece, from, to)` ?